apps/heft/src/pluginFramework/IncrementalBuildInfo.ts

```diff
@@ -0,0 +1,198 @@
+
+/**
+ * Information about an incremental build. This information is used to determine which files need to be rebuilt.
+ * @beta
+ */
+export interface IIncrementalBuildInfo {
+  /**
+   * A string that represents the configuration inputs for the build.
+   * If the configuration changes, the old build info object should be discarded.
+   */
+  configHash: string;
+
+  /**
+   * A map of absolute input file paths to their version strings.
+   * The version string should change if the file changes.
+   */
+  inputFileVersions: Map<string, string>;
+
+  /**
+   * A map of absolute output file paths to the input files they were computed from.
+   */
+  fileDependencies?: Map<string, string[]>;
+}
+
+/**
+ * Serialized version of {@link IIncrementalBuildInfo}.
+ * @beta
+ */
+export interface ISerializedIncrementalBuildInfo {
+  /**
+   * A string that represents the configuration inputs for the build.
+   * If the configuration changes, the old build info object should be discarded.
+   */
+  configHash: string;
+
+  /**
+   * A map of input files to their version strings.
+   * File paths are specified relative to the folder containing the build info file.
+   */
+  inputFileVersions: Record<string, string>;
+
+  /**
+   * Map of output file names to the input file indices used to compute them.
+   * File paths are specified relative to the folder containing the build info file.
+   */
+  fileDependencies?: Record<string, number | number[]>;
+}
+
+/**
+ * Serializes a build info object to a portable format that can be written to disk.
+ * @param state - The build info to serialize
+ * @param makePathPortable - A function that converts an absolute path to a portable path
+ * @returns The serialized build info
+ * @beta
+ */
+export function serializeBuildInfo(
+  state: IIncrementalBuildInfo,
+  makePathPortable: (absolutePath: string) => string
+): ISerializedIncrementalBuildInfo {
+  const fileIndices: Map<string, number> = new Map();
+  const inputFileVersions: Record<string, string> = {};
+
+  for (const [absolutePath, version] of state.inputFileVersions) {
+    const relativePath: string = makePathPortable(absolutePath);
+    fileIndices.set(absolutePath, fileIndices.size);
+    inputFileVersions[relativePath] = version;
+  }
+
+  const { fileDependencies: newFileDependencies } = state;
+  let fileDependencies: Record<string, number | number[]> | undefined;
+  if (newFileDependencies) {
+    fileDependencies = {};
+    for (const [absolutePath, dependencies] of newFileDependencies) {
+      const relativePath: string = makePathPortable(absolutePath);
+      const indices: number[] = [];
+      for (const dependency of dependencies) {
+        const index: number | undefined = fileIndices.get(dependency);
+        if (index === undefined) {
+          throw new Error(`Dependency not found: ${dependency}`);
+        }
+        indices.push(index);
+      }
+
+      fileDependencies[relativePath] = indices.length === 1 ? indices[0] : indices;
```

Comment:
Why not just always have this be an array? Easier to parse and serialize, and probably a trivial file size increase.

Reply:
TypeScript had done it this way, I assume they had their reasons, likely related to memory usage, but we can use a simpler model, sure.